extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,51 +14,374 @@
+import {mod} from '../../../src/utils/math';
+import {tryParseJson} from '../../../src/json';
 
 /** @const {string} */
 const EXPERIMENT = 'amp-autocomplete';
 
 /** @const {string} */
 const TAG = 'amp-autocomplete';
 
+/**
+ * Different filtering options.
+ * @const @enum {string}
+ */
+export const FilterType = {
+  SUBSTRING: 'substring',
+  PREFIX: 'prefix',
+  TOKEN_PREFIX: 'token-prefix',
+  FUZZY: 'fuzzy',
+  CUSTOM: 'custom',
+  NONE: 'none',
+};
+
 export class AmpAutocomplete extends AMP.BaseElement {
 
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = 'hello world';
+    /**
+     * The data extracted from the <script> tag optionally provided
+     * as a child. For use with static data.
+     * @private {?string}
+     */
+    this.inlineData_ = null;
+
+    /**
+     * The reference to the <input> tag provided as a child.
+     * @private {?HTMLElement}
+     */
+    this.inputElement_ = null;
+
+    /**
+     * The value of the "filter" attribute on <autocomplete>.
+     * @private {?string}
+     */
+    this.filter_ = null;
+
+    /**
+     * The value of the "min-characters" attribute on <autocomplete>.
+     * @private {?number}
+     */
+    this.minChars_ = null;
+
+    /**
+     * The value of the "max-entries" attribute on <autocomplete>.
+     * @private {?number}
+     */
+    this.maxEntries_ = null;
 
-    /** @private {?Element} */
+    /**
+     * The index of the active suggested item.
+     * @private (!number)
+     */
+    this.activeIndex_ = -1;
+
+    /**
+     * The reference to the <div> of the active suggested item.
+     * @private (!number)
+     */
+    this.activeElement_ = null;
+
+    /**
+     * The reference to the <div> that contains template-rendered children.
+     * @private {?HTMLElement}
+     */
     this.container_ = null;
   }
 
   /** @override */
   buildCallback() {
     userAssert(isExperimentOn(this.win, 'amp-autocomplete'),
         `Experiment ${EXPERIMENT} is not turned on.`);
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
+
+    this.inlineData_ = this.getInlineData_();
+
+    const inputElements = childElementsByTag(this.element, 'INPUT');
+    userAssert(inputElements.length === 1,
+        `${TAG} should contain exactly one <input> child`);
+    this.inputElement_ = inputElements[0];
+
+    this.filter_ = userAssert(this.element.getAttribute('filter'),
+        `${TAG} requires "filter" attribute.`);
+    userAssert(isEnumValue(FilterType, this.filter_),
+        `Unexpected filter: ${this.filter}`);
+
+    this.minChars_ = this.element.hasAttribute('min-characters') ?
+      parseInt(this.element.getAttribute('min-characters'), 10) : 1;
+    this.maxEntries_ = parseInt(this.element.getAttribute('max-entries'), 10);
+
+    this.container_ = this.createContainer_();
     this.element.appendChild(this.container_);
```

Comment:
We should probably meet to talk about the `this.measureMutateElement` function sometime. It's a very important method for writing performant DOM reads/writes.